atomicfu/src/appleMain/kotlin/kotlinx/atomicfu/locks/Synchronized.kt

```diff
@@ -0,0 +1,297 @@
+import platform.posix.pthread_override_qos_class_end_np
+import platform.posix.pthread_override_qos_class_start_np
+import platform.posix.pthread_override_t
+import platform.posix.pthread_self
+import platform.posix.qos_class_self
+
+import platform.posix.PTHREAD_MUTEX_ERRORCHECK
+
+
+@ThreadLocal
+var currentThreadId = 0L
+
+// Based on the compose-multiplatform-core implementation with added qos and the pool back-ported
+// from the atomicfu implementation.
+public actual open class SynchronizedObject {
+
+    companion object {
+        private const val NO_OWNER = 0L
+    }
+
+    private val owner: AtomicLong = atomic(NO_OWNER)
+    private var reEnterCount: Int = 0
+    private val waiters: AtomicInt = atomic(0)
+
+    private val monitor: DonatingMonitor by lazy { DonatingMonitor() }
+
+
+    public fun lock() {
+        var self = currentThreadId
+        if (self == 0L) {
+            currentThreadId = pthread_self().toLong()
+            self = currentThreadId
+        }
+        if (owner.value == self) {
+            reEnterCount += 1
+        } else if (waiters.incrementAndGet() > 1) {
+            waitForUnlockAndLock(self)
+        } else {
+            if (!owner.compareAndSet(NO_OWNER, self)) {
+                waitForUnlockAndLock(self)
+            }
+        }
+    }
+
+    public fun tryLock(): Boolean {
+        var self = currentThreadId
+        if (self == 0L) {
+            currentThreadId = pthread_self().toLong()
+            self = currentThreadId
+        }
+        return if (owner.value == self) {
+            reEnterCount += 1
+            true
+        } else if (waiters.incrementAndGet() == 1 && owner.compareAndSet(NO_OWNER, self)) {
+            true
+        } else {
+            waiters.decrementAndGet()
+            false
+        }
+    }
+
+
+    private fun waitForUnlockAndLock(self: Long) {
+        withMonitor(monitor) {
+            while (!owner.compareAndSet(NO_OWNER, self)) {
+                monitor.waitWithDonation(owner.value)
+            }
+        }
+    }
+
+    public fun unlock() {
+        require (owner.value == currentThreadId)
+        if (reEnterCount > 0) {
+            reEnterCount -= 1
+        } else {
+            owner.value = NO_OWNER
+            if (waiters.decrementAndGet() > 0) {
+                withMonitor(monitor) {
+                    // We expect the highest priority thread to be woken up, but this should work
+                    // in any case.
+                    monitor.nativeMutex.notify()
+                }
+            }
+        }
+    }
+
+    private inline fun withMonitor(monitor: DonatingMonitor, block: () -> Unit) {
+        monitor.nativeMutex.enter()
+        return try {
+            block()
+        } finally {
+            monitor.nativeMutex.exit()
+        }
+    }
+
+    @OptIn(kotlin.experimental.ExperimentalNativeApi::class)
+    private class DonatingMonitor {
+        val nativeMutex = mutexPool.allocate()
```

Comment:
Since mutexPool.allocate() is only called once and therefore only a single node is used. I believe we can get rid of MutexPool all together. By replacing this line for `val nativeMutex = NativeMutexNode()`. And replacing the cleaner with `val cleaner = createCleaner(nativeMutex) { it.dispose() }`.

Reply:
While reducing the number of classes and objects by one, I think this might reduce readability. I don't have a strong option either way, but I'd prefer to keep the pool aligned with the implemenations for the other platforms, at least until we have merged them in a followup?

Reply:
I would argue that having unnecessary code decreases readability, but as you said this will probably be resolved when the apple and non-apple implementations are unified.